### DISCUSSION TITLE
Will Self-hosted Supabase Be Easily Scalable?

### DISCUSSION THREAD

**Discussion by yekta** [supabase/supabase#323] [Questions] [answered]
One of the most important features of Firebase is auto-scaling with pretty much no effort on our part. I'm wondering how easy it'll be to achieve scalability when we self-host Supabase. Is there a recommended/planned strategy for easy, horizontal scaling?

**Comment by kiwicopple** [comment]
Hey @yekta, thanks for dropping in. We are testing a lot of different options, and we will open source it when we finally settle on the best option. We are starting with our hosted platform to solve scalability and performance (and of course, usability). 

Scalability is quite difficult with databases since they are stateful, but there are some promising open source "Postgres Operators" that we will be testing out.

**Reply by yekta** [reply]
Sounds promising, thanks for the quick response.

**Reply by Tragio** [reply]
@kiwicopple thank you very much. I'm with the same doubt as I'm in the stage to choose my backend 🤔 would love to know if there's something you can share about the options you're exploring or when we can find more information about it. ☺️

**Comment by kevin4dhd** [comment]
Any advance?

**Comment by PixsaOJ** [comment]
Do we have updates on this?

**Comment by inian** [comment, marked as answer]
Hi, we have a self hosting guide [here](https://supabase.io/docs/guides/self-hosting). It is a docker-compose file and scaling docker containers is hopefully not an issue. Let me know if you would like to see any more docs around this. 

We don't have a good solution yet for scaling postgres horizontally, even in the hosted version, but we are working on that.

**Reply by nosuko** [reply]
Maybe replacing postgres by cockroach DB?

**Reply by eqoram** [reply]
@inian The open source project neon (https://github.com/neondatabase/neon) might help you with scaling Postgres horizontally. :)

**Reply by techopswhiz** [reply]
@inian what about scaling just realtime horizontally?

**Comment by greendesertsnow** [comment]
It's Postgres... Sorry to put it this way but it's only as scalable as your wallet. 
There's other supabase alternatives that can horizontally and vertically scale at the same time.

**Reply by timrekelj** [reply]
So is there any information on how much concurrent users can supabase(AWS) server approximately handle? And how much can I increase this number with self-host?

**Reply by greendesertsnow** [reply]
I'll answer for Supabase. Basically this your limit. (Check the pricing...) And you can search for how much it costs on AWS.
 
Plan | Pricing | CPU | Memory | Disk IO Bandwidth | Connections: Direct | Connections: Pooler
-- | -- | -- | -- | -- | -- | --
Medium | $50 | 2-core ARM (shared) | 4 GB | Up to 2,085 Mbps | 50 (recommended) | 150 (recommended)
16XL|$3,720|64-core ARM (dedicated)|256 GB|Up to 19,000 Mbps|500 (recommended)|11,600 (recommended)


https://supabase.com/docs/guides/platform/compute-add-ons

I'm not sure if this is appropriate to write here:
You can always bring your data back to Firebase... hahaha :D
Or there's new kids on the block...
For 4gb 2CPU (which is 5 bucks on Hetzner)
Appwrite (horizontally scalable) can handle 1500 concurrent connections
Pocketbase (golang SQLite) can handle 10000 concurrent connections...
There's Parse and many others...

But postgres is a mature technology bla bla bla...

**Reply by moracabanas** [reply]
Excuse my igonrance about this topic. I was about to ask "what about scaling Postgres on k8s or k3s?"

And short answer is relational DBs are hard to scale because its nature. i.e.:   

You need sharding.

You can kind of auto-shard with citus pg extension but with some keypoints.  

I don't know how sharding and scaling is managed at the same time but I bet even if you manage to horizontal scaling your database, every time you make a replica (i.e Horizontal Pod Autoscaling), the new database needs a copy and [sharding rebalancing](https://www.citusdata.com/blog/2021/03/05/citus-10-release-open-source-rebalancer-and-columnar-for-postgres/) so write performance is 1/2 for a while until it is done.  

Another issue which comes to my mind, is when you scale any type of replicas, each of one is increasing bandwidth and it could saturate the internal network.

Anyway take a look at [citus](https://www.citusdata.com/product/community) extension, it is opensource
